include/pybind11/stl_bind.h

```diff
@@ -136,6 +136,28 @@ void vector_modifiers(enable_if_t<is_copy_constructible<typename Vector::value_t
        "Extend the list by appending all the items in the given list"
     );
 
+    cl.def("extend",
+       [](Vector &v, iterable it) {
+           const size_t old_size = v.size();
+           v.reserve(old_size + len_hint(it));
```

Comment:
I'd favour the more accurate `len` over `len_hint` if `len` is available. Something like this, perhaps:

    v.reserve(old_size + (hasattr(it, "__len__") ? len(it) : len_hint(it)));

Why am I using a `?:` here? Because, according to @yeganer it makes it clearer that this is just preallocation and that everything inside `reserve()` is just an implementation detail, compared to `if (...) { v.reserve(...); } else { v.reserve(...) }`.

I think the same should be done on line 125.

Reply:
Not necessary, the documentation for PyObject_LengthHint states that: Return an estimated length for the object o. First try to return its __actual__ length, then an estimate using __length_hint__(), and finally return the default value.

Reply:
I'm definitely missing something. Your `len_hint` will return `PyObject_LengthHint()` and won't try `PyObject_Length()`, so why is it not necessary to call `len` here first?

Reply:
PyObject_LengthHint returns the actual length if available (i.e. via a call to '__len__').  If not it tries '__length_hint__()' and failing that returns the default value (that I'm setting to 0).  This seems to be precisely the behaviour we want?

Reply:
Thanks for clearing that up. I wasn't aware that `PyObject_LengthHint()` calls `__len__()` first. Yes, this is what we want.